enhancements/network/ocp-on-smart-nics.md

```diff
@@ -0,0 +1,276 @@
+software stack on all of the NICs in the cluster. This software will
+need to be kept in sync with the software on the host PC (eg, as new
+functionality is added to OCP). The obvious answer to this problem is
+to run OpenShift on the NIC as well as on the host.
+
+The initial target of this work is the [NVIDIA BlueField-2 DPU].
+However, similar products are being worked on by other vendors which
+we will eventually want to support, so the solution should not be too
+specialized to this one product.
+
+[NVIDIA BlueField-2 DPU]: https://www.nvidia.com/en-us/networking/products/data-processing-unit/
+
+## Motivation
+
+### Goals
+
+- Extend OCP as needed to support installation on Smart NICs such as
+  the BlueField-2 where:
+
+    1. The NIC is capable of running OpenShift.
+
+    2. The NIC has an IPMI-addressable Baseboard Management Controller
+       that can be used to control and provision it independently of
+       its host.
+
+    3. The NIC implements SR-IOV in such a way that we can connect
+       pods on the host directly to an OVN bridge on the NIC.
+
+- Design an appropriate architecture for an environment containing a
+  mix of x86_64 hosts with "dumb" NICs, x86_64 hosts with Smart NICs,
+  and the ARM-based Smart NICs themselves. (As discussed below, such
+  an environment may involve one or more OCP clusters.)
+
+- Define how OCP installs and upgrades will work in such an
+  environment.
+
+This enhancement does not cover what we actually _do_ with the NICs
+after they are provisioned. The primary use case of OVN offload is
+covered by "[Smart NIC OVN Offload]". It is possible that we may also
+want to allow customers to move additional network, security, and
+monitoring functionality from the CPU to the DPU in the future.
+
+[Smart NIC OVN Offload]: https://github.com/openshift/enhancements/pull/732
+
+### Non-Goals
+
+- Support for ASIC- or FPGA-based Smart NICs.
+
+- Support for ARM-based Smart NICs where we would not be able to run
+  RHCOS on the NIC (eg, because the NIC requires proprietary drivers).
+
+## Design Details
+
+### Clusters Will Contain a Mix of Smart and "Dumb" NICs
+
+OVN Kubernetes is not currently capable of mixing offloaded and
+non-offloaded networking on the same node, so every pod on a node
+containing a Smart NIC must have offloaded networking, which, in the
+current architecture, means it must explicitly _request_ offloaded
+networking. In particular, this means that most pods created by
+OpenShift operators cannot be deployed on a node with a Smart NIC, and
+so all masters and at least a few workers will need to have "dumb"
+NICs.
+
+(While the exact architectural constraints may change in the future,
+it is likely that some customers will always want the possibility of
+only having Smart NICs on a subset of nodes.)
+
+### Overall Cluster Architecture
+
+There are several possible overall architectures:
+
+  - The Smart NICs could be nodes within the same OCP cluster as their
```

Comment:
is there a motivation for this topology?  in prior discussions, this broke one of the intended benefits of ensuring strong security isolation between infrastructure and workload.

Reply:
BlueField2 supports three different modes, with each has different level of security isolation.
The default mode is not the strongest isolation mode, I assume we may start from the default mode and then allow switching of the modes as user may have different requests on the security isolation.

Reply:
The motivation is "it's way easier", so if there are customers interested in the increased network throughput who don't care as much about the isolation stuff, then it may be a good approach as a first version. If there aren't enough customers who would be interested in that, then it wouldn't make sense.

(In a two-cluster architecture we need to either have one ovn-kubernetes master programming two sets of OVN databases, or else two ovn-kubernetes masters, one of which talks to two different apiservers. Either way, there's going to be a bunch of rearchitecting.)

Reply:
@derekwaynecarr all our stuff is containerized already; the same stuff that runs on the host today (OVS, ovn-controller, ovnkube, network metrics, etc) moves to the NIC.  It makes little sense to write a container lifecycle manager since we already have one: kubelet. 

For the security angle, it may not be worth having a completely separate control plane just for SmartNIC control, unless you're already going to have that for other reasons. The point is that the hosts are untrusted and since all networking through the NIC, workloads that somehow compromise the host can be detected, monitored, and isolated (I guess?). The NIC wouldn't run regular workloads, just control-plane stuff.